src/mame/tandberg/tdv2100_kbd.cpp

```diff
@@ -0,0 +1,555 @@
+	m_keyparams(*this, "keyboard_params_rom"),
+	m_matrix(*this, "X%u", 0),
+	m_scan_clock(*this, "scan_clock"),
+	m_sw_all_cap(*this, "sw_all_cap"),
+	m_online_led(*this, "led_online"),
+	m_carrier_led(*this, "led_carrier"),
+	m_error_led(*this, "led_error"),
+	m_enquiry_led(*this, "led_enquiry"),
+	m_ack_led(*this, "led_ack"),
+	m_nak_led(*this, "led_nak"),
+	m_wait_led(*this, "led_wait"),
+	m_shiftlock_led(*this, "led_shiftlock"),
+	m_write_kstr_cb(*this),
+	m_write_cleark_cb(*this),
+	m_write_linek_cb(*this),
+	m_write_transk_cb(*this),
+	m_write_break_cb(*this)
+{
+	column_counter = 0;
+	shift = false;
+	shift_lock = false;
+	control = false;
+	for(int i=0; i<15; i++)
+	{
+		keystate[i] = 0;
+	}
+
+	inhibit_key_from_params = false;    // Set by jumper 40-42
+
+	char_buffer = 0x00;
+	key_nr_in_buffer = 0xFF;
+	key_repeat_delay_ms = 800;          // Set by potentiometer, 500ms -> 1250ms
+	key_repeat_rate_hz = 20;            // Set by potentiometer, 10Hz -> 30Hz
+}
+
+///////////////////////////////////////////////////////////////////////////////
+//
+// General machine state
+//
+
+void tandberg_tdv2100_keyboard_device::device_start()
+{
+	m_key_repeat_trigger = timer_alloc(FUNC(tandberg_tdv2100_keyboard_device::key_repeat), this);
+	m_online_led.resolve();
+	m_carrier_led.resolve();
+	m_error_led.resolve();
+	m_enquiry_led.resolve();
+	m_ack_led.resolve();
+	m_nak_led.resolve();
+	m_wait_led.resolve();
+	m_shiftlock_led.resolve();
+}
+
+void tandberg_tdv2100_keyboard_device::device_reset()
+{}
+
+///////////////////////////////////////////////////////////////////////////////
+//
+// Key Scanning
+//
+
+void tandberg_tdv2100_keyboard_device::scan_next_column(int state)
+{
+	if(state)
+	{
+		int column = m_matrix[column_counter]->read();
+
+		for(int row_counter=0; row_counter<8; row_counter++)
+		{
+			int key_mask = 1<<row_counter;
+			int key_state = column&key_mask;
+			if(key_state != (keystate[column_counter]&key_mask))
+			{
+			// This simulates the enumeration done by the internal LUT of the 20-04592-013
+			uint8_t key_nr = (~((column_counter-1)*8 + row_counter + 1))&0xFF;
+				if(key_state)
+				{
+					switch(column_counter)
+					{
+						case 0:
+							switch(row_counter)
+							{
+								case 0:
+								case 1:
+									shift = true;
+									break;
+
+								case 2:
+									shift_lock = !shift_lock;
+									m_shiftlock_led = !shift_lock;
+									break;
+
+								case 3:
+									control = true;
+									break;
+							}
+							break;
+
+						case 14:
+							switch(row_counter)
+							{
+								case 4:
+									m_write_cleark_cb(1);
+									break;
+
+								case 5:
+									m_write_transk_cb(1);
+									break;
+
+								case 6:
+									m_write_linek_cb(1);
+									break;
+
+								case 7:
+									m_write_break_cb(1);
+									break;
+							}
+							break;
+
+						default:
+							new_keystroke(key_nr, (shift||shift_lock), control);
+					}
+					keystate[column_counter] |= key_mask;
```

Comment:
You’re better off just updating this to `column` after the loop rather than changing one bit at a time.